clients/iceberg-extensions/src/main/scala/org/apache/spark/sql/execution/datasources/v2/NessieUtils.scala

```diff
@@ -0,0 +1,205 @@
+/*
+ * Copyright (C) 2020 Dremio
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.sql.execution.datasources.v2
+import org.apache.spark.sql.connector.catalog.CatalogPlugin
+import org.apache.spark.sql.util.CaseInsensitiveStringMap
+import org.apache.spark.sql.{AnalysisException, SparkSession}
+import org.projectnessie.api.params.CommitLogParams
+import org.projectnessie.client.{NessieClient, StreamingUtil}
+import org.projectnessie.model.{Branch, CommitMeta, Hash, Reference}
+
+import java.time.{LocalDateTime, ZoneId}
+import scala.collection.JavaConverters._
+
+object NessieUtils {
+  def calculateRef(
+      branch: String,
+      ts: Option[String],
+      nessieClient: NessieClient
+  ): Reference = {
+    //todo we are assuming always in UTC. ignoring tz set by spark etc
+    val timestamp = ts
+      .map(x => x.replaceAll("`", ""))
+      .map(x => LocalDateTime.parse(x).atZone(ZoneId.of("UTC")).toInstant)
+      .orNull
+    if (timestamp == null) {
+      nessieClient.getTreeApi.getReferenceByName(branch)
+    } else {
+      var last: Option[CommitMeta] = None
+      // todo this is icky
+      for (cm <- StreamingUtil
+             .getCommitLogStream(
+               nessieClient.getTreeApi,
+               branch,
+               CommitLogParams.empty()
+             )
+             .iterator
+             .asScala) {
+        if (cm.getCommitTime.isBefore(timestamp)) {
+          Hash.of(cm.getHash)
+        }
+        last = Some(cm)
+      }
+      last
+        .map(x => Hash.of(x.getHash))
+        .getOrElse(nessieClient.getTreeApi.getReferenceByName(branch))
+    }
+  }
+
+  def nessieClient(
+      currentCatalog: CatalogPlugin,
+      catalog: Option[String]
+  ): NessieClient = {
+    val catalogName = catalog.getOrElse(currentCatalog.name)
+    val catalogConf = SparkSession.active.sparkContext.conf
+      .getAllWithPrefix(s"spark.sql.catalog.$catalogName.")
+      .toMap
+    NessieClient
```

Comment:
This worries me.
`NessieClient` is `AutoCloseable`, but it's never closed. Maybe just refactor the `lazy val` usage.

Reply:
agreed. Refactored